quota/mysql/mysql_quota_test.go

```diff
@@ -0,0 +1,364 @@
+			specs:              []quota.Spec{{Group: quota.Global, Kind: quota.Write}},
+		},
+		{
+			desc:               "globalWriteMultiToken",
+			unsequencedRows:    10,
+			maxUnsequencedRows: 20,
+			numTokens:          5,
+			specs:              []quota.Spec{{Group: quota.Global, Kind: quota.Write}},
+		},
+		{
+			desc:               "globalWriteOverQuota1",
+			unsequencedRows:    20,
+			maxUnsequencedRows: 20,
+			numTokens:          1,
+			specs:              []quota.Spec{{Group: quota.Global, Kind: quota.Write}},
+			wantErr:            true,
+		},
+		{
+			desc:               "globalWriteOverQuota2",
+			unsequencedRows:    15,
+			maxUnsequencedRows: 20,
+			numTokens:          10,
+			specs:              []quota.Spec{{Group: quota.Global, Kind: quota.Write}},
+			wantErr:            true,
+		},
+		{
+			desc:      "unlimitedQuotas",
+			numTokens: 10,
+			specs: []quota.Spec{
+				{Group: quota.User, Kind: quota.Read, User: user},
+				{Group: quota.Tree, Kind: quota.Read, TreeID: tree.TreeId},
+				{Group: quota.Global, Kind: quota.Read},
+				{Group: quota.User, Kind: quota.Write, User: user},
+				{Group: quota.Tree, Kind: quota.Write, TreeID: tree.TreeId},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		if err := setUnsequencedRows(ctx, db, tree, test.unsequencedRows); err != nil {
+			t.Errorf("setUnsequencedRows() returned err = %v", err)
+			continue
+		}
+
+		// Test general cases using select count(*) to avoid flakiness / allow for more
+		// precise assertions.
+		// See TestQuotaManager_GetTokens_InformationSchema for information schema tests.
+		qm := &mysqlq.QuotaManager{DB: db, MaxUnsequencedRows: test.maxUnsequencedRows, UseSelectCount: true}
+		err := qm.GetTokens(ctx, test.numTokens, test.specs)
+		if hasErr := err != nil; hasErr != test.wantErr {
+			t.Errorf("%v: GetTokens() returned err = %q, wantErr = %v", test.desc, err, test.wantErr)
+		}
+	}
+}
+
+func TestQuotaManager_GetTokens_InformationSchema(t *testing.T) {
+	ctx := context.Background()
+
+	maxUnsequenced := 20
+	globalWriteSpec := []quota.Spec{{Group: quota.Global, Kind: quota.Write}}
+
+	// Make both variants go through the test.
+	tests := []struct {
+		useSelectCount bool
+	}{
+		{useSelectCount: true},
+		{useSelectCount: false},
+	}
+	for _, test := range tests {
+		desc := fmt.Sprintf("useSelectCount = %v", test.useSelectCount)
+
+		db, err := integration.GetTestDB(fmt.Sprintf("GetTokensInformationSchemaTest_%v", test.useSelectCount))
+		if err != nil {
+			t.Errorf("%v: GetTestDB() returned err = %v", desc, err)
+			continue
+		}
+
+		tree, err := createTree(ctx, db)
+		if err != nil {
+			t.Errorf("%v: createTree() returned err = %v", desc, err)
+			continue
+		}
+
+		qm := &mysqlq.QuotaManager{DB: db, MaxUnsequencedRows: maxUnsequenced, UseSelectCount: test.useSelectCount}
+
+		// All GetTokens() calls where leaves < maxUnsequenced should succeed:
+		// information_schema may be outdated, but it should still point to a valid point in
```

Comment:
"point to a valid point in the past" reads a bit unclear, please reword. We also don't know how out of date this could be in practice.